renderdoc/driver/vulkan/vk_acceleration_structure.cpp

```diff
@@ -580,183 +613,265 @@ bool VulkanAccelerationStructureManager::Serialise(SerialiserType &ser, Resource
                                                    CaptureState state)
 {
   VkDevice d = !IsStructuredExporting(state) ? m_pDriver->GetDev() : VK_NULL_HANDLE;
-  const bool replayingAndReading = ser.IsReading() && IsReplayMode(state);
   VkResult vkr = VK_SUCCESS;
 
   byte *contents = NULL;
-  uint64_t contentsSize = initial ? initial->mem.size : 0;
-  MemoryAllocation mappedMem;
-
-  // Serialise this separately so that it can be used on reading to prepare the upload memory
-  SERIALISE_ELEMENT(contentsSize);
-
-  const VkDeviceSize nonCoherentAtomSize = m_pDriver->GetDeviceProps().limits.nonCoherentAtomSize;
-
-  // the memory/buffer that we allocated on read, to upload the initial contents.
-  MemoryAllocation uploadMemory;
-  VkBuffer uploadBuf = VK_NULL_HANDLE;
 
   if(ser.IsWriting())
   {
-    if(initial && initial->mem.mem != VK_NULL_HANDLE)
-    {
-      const VkDeviceSize size = AlignUp(initial->mem.size, nonCoherentAtomSize);
+    VkAccelerationStructureInfo *asInfo = initial->accelerationStructureInfo;
+    RDCASSERT(asInfo != NULL);
+    SERIALISE_ELEMENT(*asInfo).Hidden();
```

Comment:
Sorry I should have clarified in my earlier comment - anything which is serialising should be outside of any reading/writing control flow to ensure that the serialisation is easily readable as the same in both directions. The flow of a serialisation function is as below:

```
Data data;
if(IsWriting())
{
  // do anything to prepare data
}

SERIALISE(data);

if(IsReading())
{
process(data);
}
```

For complex cases like initial states here when we're uploading data directly you sometimes need a codepath before the serialisation for reading too, to map the buffer etc. Or you might need to do "two passes" of serialisation, one for the metadata, then allocate a buffer, then serialise the data. In any case the actual call to serialise should be kept outside of control flow.

This is a tradeoff between making it super clear that the serialisation won't mismatch, and the simplification possible by abstracting out as you had before with separate functions.